lib/parse_identifier.rb

```diff
@@ -0,0 +1,121 @@
+require 'identifiers'
+
+# https://github.com/altmetric/identifiers
+#
+# Identifiers::DOI.extract('example: 10.123/abcd.efghi')
+# # => ["10.123/abcd.efghi"]
+#
+# It can support many identifiers:
+# Identifiers::AdsBibcode.extract('')
+# Identifiers::ArxivId.extract('')
+# Identifiers::DOI.extract('')
+# Identifiers::Handle.extract('')
+# Identifiers::ISBN.extract('')
+# Identifiers::NationalClinicalTrialId.extract('')
+# Identifiers::ORCID.extract('')
+# Identifiers::PubmedId.extract('')
+# Identifiers::RepecId.extract('')
+# Identifiers::URN.extract('')
+
+# Exception for incorrect instantiation of a parser (not necessarily bad data)
+class ParseIdentifierTypeError < StandardError
+
+end
+
+# Exception for bad data
+class ParseIdentifierInvalidError < StandardError
+
+end
+
+# Parse the identifiers in a PublicationIdentifier
+#
+# This is a base class for ParseIdentifier* subclasses.
+#
+# This class should try to work with the input parameter as a read-only object.
+# It should only modify it when explicitly asked to `update` it.  If it's
+# an active record object, assign new values to it, but don't save it, leave
+# that persistence action to the consumer; see the runner script in
+# @see script/publication_identifier_normalization.rb
+class ParseIdentifier
+
+  URI_PREFIX = nil # subclasses override this constant
+
+  attr_reader :pub_id
+  attr_reader :type
+
+  # @param pub_id [PublicationIdentifier]
+  def initialize(pub_id)
+    @pub_id = pub_id
+    @type = pub_id[:identifier_type]
+    raise(ParseIdentifierTypeError, "INVALID TYPE #{pub_id.inspect}") unless match_type
+    raise(ParseIdentifierInvalidError, "EMPTY DATA #{pub_id.inspect}") if empty?
+    raise(ParseIdentifierInvalidError, "INVALID DATA #{pub_id.inspect}") unless valid?
+  end
+
+  def empty?
+    pub_id[:identifier_value].blank? && pub_id[:identifier_uri].blank?
+  end
+
+  # Update the pub_id with parsed data
+  # @return pub_id [PublicationIdentifier]
+  def update
+    pub_id[:identifier_value] = value
+    pub_id[:identifier_uri] = uri
+    pub_id
+  end
+
+  # Does the data validate?
+  def valid?
+    # the base class does no validations
+    true
```

Comment:
Seems like `false` would be a safer default, such that any class extending this wouldn't be silently reporting values as valid when no actual check is happening.

Reply:
The base class can check for empty data (which is does first in the initializer and it raises an exception for empty data).  It has no capacity for validation (normalization) and, by design, it should just return whatever it is given.  This ensures that if it is given any `SulPubId` or `PublicationItemID` (or whatever else _might_ get thrown in there), that it will only check for blank data and otherwise leave it alone.  Only the subclasses can validate/normalize data and they override this method.  Consider the distinct functions of the `blank?` vs. `valid?` methods.

Reply:
But an object can be fully initialized and then have the value set to `nil` subsequently.  At that point, `valid?` should not be returning `true`, right?

Reply:
Well, that depends on the sub-class, but essentially that's the idea.  The ISBN is allowed to have a `nil` URI, but otherwise everything should have a `value` and a `uri`.  If the parser cannot extract a `value`, the thing should not be `valid?` any more.  However, note that the intializer calls the `valid?` method and it calls the `value` method, which will run the parser.  So, in effect, the initializer does all the validation required and raises an exception at that point.  Therefore, the possibility of subsequent changes is eliminated (e.g. it can never transition from invalid to valid).  Let's discuss if it still doesn't add up for some reason.

Reply:
I suggest if you checkout the branch and try some of the things you want, you will see what happens.